src/Eio/Adapter.php

```diff
@@ -309,6 +309,56 @@ public function close($fd)
         });
     }
 
+    /**
+     * Reads the entire file.
+     *
+     * This is an optimization for adapters which can optimize
+     * the open -> write -> close sequence into one call.
+     *
+     * @param string $path
+     * @param int $offset
+     * @param int|null $length
+     * @return PromiseInterface
+     */
+    public function getContents($path, $offset = 0, $length = null)
+    {
+        if ($length === null) {
+            return $this->stat($path)->then(function ($stat) use ($path, $offset) {
+                return $this->getContents($path, $offset, $stat['size']);
+            });
+        }
+
+        return $this->open($path, 'r')->then(function ($fd) use ($offset, $length) {
+            return $this->read($fd, $length, $offset)->always(function () use ($fd) {
+                return $this->close($fd);
+            });
+        });
+    }
+
+    /**
+     * Writes directly to the file.
+     *
+     * This is an optimization for adapters which can optimize
+     * the open -> write -> close sequence into one call.
+     *
+     * Supported flags are `LOCK_EX` and `FILE_APPEND`.
+     *
+     * @param string $path
+     * @param string $content
+     * @param int $flags Flags for writing the contents.
+     * @return PromiseInterface
+     */
+    public function putContents($path, $content, $flags = 0)
+    {
+        $oflags = ($flags & FILE_APPEND ? 'cwa' : 'cw').($flags & LOCK_EX ? 'e' : '');
```

Comment:
Can you confirm the `LOCK_EX` works as expected? It's my understanding the `e` flag maps to `EIO_O_EXCL` which is essentially means this call should "fail if the file already exists", whereas `LOCK_EX` would acquire an exclusive lock (so no other process can read from it while it is being written to).

I'm not intimately familar with libeio, but if we don't have a mapping for this flag, we might as well leave these flags out from the interface for now? If so, my vote would be to have a dedicated `append()` method instead of passing this as a flag :+1:

Reply:
@clue Yes, you're right about the `EIO_O_EXCL` flag. I misinterpreted the description. The description of the fcntl.h page is a lot clearer.

Eio effectively does not support file locking (libeio does support memory locking, don't ask me why). But that does not mean we should exclude this for other adapters. I think instead we should document this as a specific limitation by Eio and if locking is needed a different adapter should be used.

Appending can effectively be achieved by using `open` with `a` flag (`wa` for eio) and using `write` for the tradional "stream" way. In terms of this method it's a flag for the included adapters. I don't think we should separate this further as it's effectively the same underlying call but with different flags (in different places) and we should treat it the same way.